### ISSUE TITLE
How to enable CUDA on Kaldi in transcrbing step

### ISSUE THREAD

**Issue by lightsailpro** [lowerquality/gentle#148]
I have a K40c GPU configured. And in ext/install_kaldi.sh file, I changed use_cuda=yes, run the install_kaldi.sh, then restart the gentle server. But I GPU is still not being used during the trancription step.
Does Gentle support the GPU in speech recognition step? Thanks.

/configure --static --static-math=yes --static-fst=yes --use-cuda=yes

**Comment by strob** [comment]
No, I don't believe Gentle supports GPU-based decoding. Check Kaldi docs
for more information.

**Comment by lightsailpro** [comment]
Forgot to run make after the change in install_kaldi.sh. 
./configure --static --static-math=yes --static-fst=yes --use-cuda=yes --cudatk-dir=/usr/local/cuda

but "make" generated the following error:
 
/home/ml/gentle/ext/kaldi/src/cudamatrix/../cudamatrix/cu-array-inl.h:141: undefined reference to `cudaMemcpy'

**Comment by adamelmore** [comment]
I started to make a new issue, but thought I'd just revive the conversation here.

@strob not sure if you're aware of it, but the Kaldi team is currently working on a GPU decoder. [There's a paper](http://www.danielpovey.com/files/2018_interspeech_gpu_wfst.pdf) and a [WIP PR](https://github.com/kaldi-asr/kaldi/pull/2322) with lots of ongoing conversation. Not entirely sure what the timetable is for this to be merged, but I've been watching pretty closely as I have a need for ultra-fast lattice generation. We're currently using gentle in our project, and I'm open to working on a PR here once this stuff is merged on the Kaldi side. Just wanted to get your thoughts, first, and see if this was on your radar.

**Comment by yondu22** [comment]
Does anyone know performance improvement with this?
https://medium.com/voicetube/build-gentle-w-cuda-enabled-kaldi-cb9eac86afc3

**Comment by strob** [comment]
I don’t believe there are significant performance improvements decode-side but would love to see a comparison!

**Comment by oerdem19** [comment]
I have done all these but it seems that GPU is still not being utilized.

**Comment by trias702** [comment]
I have successfully compiled the Gentle k3 binary against Kaldi master branch with CUDA. Unfortunately, I agree with oerdem19 that the GPU is not actually being utilised by k3. The GPU VRAM is being used, so some kind of payload is going to the GPU, but no actual calcuations are happening on the GPU.

Looking at the k3.cc source, it makes sense that nothing is happening because the CuDevice object is never used:

```
CuDevice &cu_device = CuDevice::Instantiate();
cu_device.SetVerbose(true);
cu_device.SelectGpuId("yes");
```

However, if you look at the rest of the k3 code, that `cu_device` object is never used anywhere, which I think is the problem. It should be passed to some decoding function but it isn't.

Has anyone discovered any way of potentially getting the decoding to happen on GPU?

**Comment by Lorenzoncina** [comment]
Any update on this topic?
